Code/MinimalLib/cffiwrapper.cpp

```diff
@@ -0,0 +1,593 @@
+#include <cstring>
+#include <iostream>
+
+#include <RDGeneral/versions.h>
+#include <atomic>
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/MolPickler.h>
+#include <GraphMol/SmilesParse/SmilesParse.h>
+#include <GraphMol/SmilesParse/SmilesWrite.h>
+#include <GraphMol/SmilesParse/SmartsWrite.h>
+#include <GraphMol/FileParsers/FileParsers.h>
+#include <GraphMol/MolDraw2D/MolDraw2D.h>
+#include <GraphMol/MolDraw2D/MolDraw2DSVG.h>
+#include <GraphMol/MolDraw2D/MolDraw2DUtils.h>
+#include <GraphMol/MolInterchange/MolInterchange.h>
+#include <GraphMol/Substruct/SubstructMatch.h>
+#include <GraphMol/Descriptors/Property.h>
+#include <GraphMol/Descriptors/MolDescriptors.h>
+#include <GraphMol/Fingerprints/MorganFingerprints.h>
+#include <GraphMol/Fingerprints/Fingerprints.h>
+#include <GraphMol/Depictor/RDDepictor.h>
+#include <GraphMol/CIPLabeler/CIPLabeler.h>
+#include <GraphMol/Abbreviations/Abbreviations.h>
+#include <GraphMol/DistGeomHelpers/Embedder.h>
+#include <DataStructs/BitOps.h>
+
+#include "common.h"
+
+#include <sstream>
+#include <RDGeneral/BoostStartInclude.h>
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <RDGeneral/BoostEndInclude.h>
+
+#include <INCHI-API/inchi.h>
+
+#include <rapidjson/document.h>
+#include <rapidjson/stringbuffer.h>
+#include <rapidjson/writer.h>
+
+namespace rj = rapidjson;
+
+using namespace RDKit;
+
+#if (defined(__GNUC__) || defined(__GNUG__))
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wconversion-null"
+#endif
+
+namespace {
+char *str_to_c(const std::string &str, size_t *len = nullptr) {
+  if (len) {
+    *len = str.size();
+  }
+  char *res;
+  res = (char *)malloc(str.size() + 1);
+  memcpy((void *)res, (const void *)str.c_str(), str.size());
+  res[str.size()] = 0;
+  return res;
+}
+char *str_to_c(const char *str) {
+  char *res;
+  res = (char *)malloc(strlen(str) + 1);
+  strcpy(res, str);
+  return res;
+}
+}  // namespace
+
+#define MOL_TO_PKL(mol, mol_pkl, mol_pkl_sz)                                   \
+  {                                                                            \
+    unsigned int propFlags = PicklerOps::PropertyPickleOptions::AllProps ^     \
+                             PicklerOps::PropertyPickleOptions::ComputedProps; \
+    std::string pkl;                                                           \
+    MolPickler::pickleMol(mol, pkl, propFlags);                                \
+    free(*mol_pkl);                                                            \
+    *mol_pkl = str_to_c(pkl, mol_pkl_sz);                                      \
+  }
+
+#define MOL_FROM_PKL(mol, pkl, pkl_sz)       \
+  if (!(pkl) || !(pkl_sz)) {                 \
+    return NULL;                             \
+  }                                          \
+  std::string mol##inp_str((pkl), (pkl_sz)); \
+  RWMol mol(mol##inp_str);                   \
+  mol.setProp(common_properties::_StereochemDone, 1, true);

```

Comment:
I'd prefer functions for these instead of #defines?

Reply:
yeah, I had that "I shouldn't be using #defines" guilty feeling...

fixed now

Reply:
@ricrogz : it would be cool if you could take a quick look at that last commit to see if you agree with the way I removed the #defines.

Reply:
Done, sorry for the delay :)